File: C++/169. Majority Element.cpp
#include <bits/stdc++.h>

using namespace std;

int majorityElement(vector<int> &v)
{
    unordered_map<int, int> m;

    for (int i = 0; i < v.size(); i++)
        m[v[i]]++;

    int major = v[0];
    int count = m[v[0]];

    for (auto i : m)
    {
        if (count <= i.second)
        {
            count = i.second;
            major = i.first;
        }
    }

    return major;
}

/*
    Above Approach have Space Complexity O(n)
    we have to optimize it

sort and return the middle element
Time: O(nLogn)
space: O(1)

*/

int majorityElement(vector<int> &nums)
{
    sort(nums.begin(), nums.end());
    return nums[nums.size() / 2];
}

/*
    Best Solution using moor's voting Alg
*/
int majorityElement(vector<int> &nums)
{
    int count = 1;
    int major = nums[0];

    for (int i = 1; i < nums.size(); i++)
    {
        if (count == 0)
            major = nums[i];

        if (major == nums[i])
            count++;
        else
            count--;
    }

    return major;
}

File: Java/169. Majority Element.java
import java.util.Arrays;
import java.util.HashMap;

/* 
 * Time: O(nlogn)
 * space: O(1)
 */
class Solution {

    public int majorityElement_1(int[] nums) {
        Arrays.sort(nums);
        return nums[nums.length / 2];
    }

    /*
     * HashMap Solution
     * Time: O(n)
     * space: O(n)
     */

    public int majorityElement_2(int[] nums) {

        var map = new HashMap<Integer, Integer>();
        int major = 0;
        int count = 0;

        for (var ele : nums) {
            map.put(ele, map.getOrDefault(ele, 0) + 1);

            if (map.get(ele) > count) {
                major = ele;
                count = map.get(ele);
            }
        }

        return major;
    }

    /*
     * Optimized Approach
     * moor's voting Alg
     * 
     * Time: O(n)
     * space: O(1)
     */
    public int majorityElement_3(int[] nums) {
        int count = 0;
        int major = 0;

        for (int num : nums) {
            if (count == 0)
                major = num;

            if (major == num)
                count++;
            else
                count--;
        }

        return major;
    }
}
